### PR TITLE
Reset Password bug fixes / enhancements

### PR DESCRIPTION
**Pull request by nickretallack**
Fixes https://github.com/mattupstate/flask-security/issues/375 and makes the reset token actually expire once it is used.

### CODE REVIEW

**Comment by nickretallack-business** [comment]
Can this be accepted added soon?

**Comment by nickretallack** [comment]
Can this be accepted soon?

**Comment by mattupstate** [comment]
@nickretallack I'm not sure I understand this so-called enhancement. Nor do I see any tests that illustrate the use case.

**Comment by nickretallack-business** [comment]
Details are in the linked issue.  I'll look into how to test it.

The primary issue is, if you attempt to log in as a user who has no password set, you'll raise an exception.

**Comment by nickretallack** [comment]
Details are in the linked issue.  I'll look into how to test it.

The primary issue is, if you attempt to log in as a user who has no password set, you'll raise an exception.

**Comment by jonafato** [comment]
It seems to me that this pull request addresses two unrelated issues.

1. Currently, the MD5 has of a user's current password is used in the generation of a password reset token. When a user has no password (to pull some reasons from the linked issue: a user only currently authenticated via some social API or an account created on behalf of some user as part of an invite-only system), the password can be `None` causing the MD5 step to raise an exception. I agree that this is an issue that needs to be solved.

2. Additionally, the PR is adding explicit logic to consider a password reset token to be invalid if the MD5 of the user's current password hash is not equal to the value pulled from the token. I should note that this is a breaking change if anyone is relying tokens working more than once within their max_age lifetime. This is not a good thing to depend on, but it should make its way into the changelog if accepted.

Both of these changes seem like welcome additions, but they should probably be submitted as separate changes and should definitely be accompanied by tests.

**Comment by mattupstate** [comment]
Users without passwords is strange to me. If a user is created without a password then it is most likely being done outside the context of Flask-Security code. If this is necessary, why would one not create a randomly generated password for the user when the user is created to avoid all this?

**Comment by mattupstate** [comment]
@nickretallack you're going to have to forgive me for being so short sighted. In all honesty, I always thought this was the way I had implemented it. Clearly I'm a moron. What you've contributed here is what I always intended to have, so let me thank you for the additions and again apologize for being such a pain about your PR.

@jonafato do you know of anyone that expects password reset tokens to work multiple times?

**Comment by jonafato** [comment]
I meant to post the following earlier today but closed the tab.

@mattupstate A concrete example: for PyGotham, we occasionally create accounts on behalf of invited speakers. In the interest of not sending a randomly generated password in cleartext, we send them a password reset link. Our current workaround for the `MD5(None)` issue is to set the password hash to some value that a legitimate hash will never collide with and generate the reset link based on that. This is roughly equivalent to what you suggest, but it would be convenient if it was not a required step.

The other change ensures that a password reset link cannot be used after the hash with which it was generated has changed. This is not currently the case, and I think it's a necessary step. If someone does currently expect a password reset token to work more than once, they're explicitly weakening the security of their application and should be discouraged / prevented from doing so. That said, it's an important change in behavior, and my main point was that this should make it into the changelog.

**Comment by mattupstate** [comment]
Understood! Now we just need some tests.

File: flask_security/recoverable.py [hunk 1]
```diff
@@ -62,11 +63,17 @@ def reset_password_token_status(token):
...
-        expired, invalid, user = reset_password_token_status('...')
+        expired, invalid, user, data = reset_password_token_status('...')
 
     :param token: The password reset token
     """
-    return get_token_status(token, 'reset', 'RESET_PASSWORD')
+    expired, invalid, user, data = get_token_status(token, 'reset', 'RESET_PASSWORD', return_data=True)
+    if not invalid:
+        password_hash = md5(user.password) if user.password else None
+        if password_hash != data[1]:
```
Comment (mattupstate) [inline flask_security/recoverable.py]: should be using `werkzeug.security.safe_str_cmp` here.

**Comment by nickretallack** [comment]
New pull request: https://github.com/mattupstate/flask-security/pull/389#issuecomment-101155709

**Pull request closed without merging**